sdk/eventhub/azure-eventhubs/azure/eventhub/_consumer.py

```diff
@@ -0,0 +1,163 @@
+        link_property_timeout_ms = (self._client._config.receive_timeout or self._timeout) * 1000  # pylint:disable=protected-access
+        self._link_properties[types.AMQPSymbol(TIMEOUT_SYMBOL)] = types.AMQPLong(int(link_property_timeout_ms))
+        self._handler = None
+        self._track_last_enqueued_event_properties = track_last_enqueued_event_properties
+        self._last_enqueued_event_properties = {}
+        self._last_received_event = None
+
+    def _create_handler(self):
+        source = Source(self._source)
+        if self._offset is not None:
+            source.set_filter(self._offset._selector())  # pylint:disable=protected-access
+        desired_capabilities = None
+        if self._track_last_enqueued_event_properties:
+            symbol_array = [types.AMQPSymbol(RECEIVER_RUNTIME_METRIC_SYMBOL)]
+            desired_capabilities = utils.data_factory(types.AMQPArray(symbol_array))
+
+        properties = create_properties(self._client._config.user_agent)  # pylint:disable=protected-access
+        self._handler = ReceiveClient(
+            source,
+            auth=self._client._create_auth(),  # pylint:disable=protected-access
+            debug=self._client._config.network_tracing,  # pylint:disable=protected-access
+            prefetch=self._prefetch,
+            link_properties=self._link_properties,
+            timeout=self._timeout,
+            error_policy=self._retry_policy,
+            keep_alive_interval=self._keep_alive,
+            client_name=self._name,
+            receive_settle_mode=uamqp.constants.ReceiverSettleMode.ReceiveAndDelete,
+            auto_complete=False,
+            properties=properties,
+            desired_capabilities=desired_capabilities)
+
+        self._handler._streaming_receive = True  # pylint:disable=protected-access
+        self._handler._message_received_callback = self._message_received  # pylint:disable=protected-access
+
+    def _open_with_retry(self):
+        return self._do_retryable_operation(self._open, operation_need_param=False)
+
+    def _message_received(self, message):
+        # pylint:disable=protected-access
+        event_data = EventData._from_message(message)
+        trace_link_message(event_data)
+        self._last_received_event = event_data
+        self._on_event_received(event_data)
+
+    def receive(self):
+        retried_times = 0
+        last_exception = None
+
+        while retried_times < self._client._config.max_retries:  # pylint:disable=protected-access
+            try:
+                self._open()
+                self._handler.do_work()
+                return
+            except uamqp.errors.LinkDetach as ld_error:
+                if ld_error.condition == uamqp.constants.ErrorCodes.LinkStolen:
+                    raise self._handle_exception(ld_error)
+            except Exception as exception:  # pylint: disable=broad-except
+                if not self.running:  # exit by close
```

Comment:
Can I get some context here? Seems weird to me to catch a general exception, and return on it just by the value of a local attribute.

Reply:
@yunhaoling - Do you have some insight here?

If I'm following this correctly, `self.running` will be false if an error occured that required the AMQP handler and/or connection to be shutdown. However this doesn't mean that it's not a retryable error - so shouldn't we still be retrying in this case?

Reply:
+ @YijunXieMS

Reply:
so the reason I put it here is because:

1. The load balance is working in another thread which is responsible for cancelling task (close consumer).
2.  The receive runs in main thread
3. 1 and 2 happens simultaneously, the `receive` could get error because the handler is closed by the load balance thread. In that case, this exception is expected.

Reply:
To me this begs the question of if reading from a closed handler results in a reliable exception that we could catch precisely instead of a generic Exception?

Reply:
@KieranBrantnerMagee taking your advice, so that check would be:                
```python
if not self.running and isinstance(
    exception,
    (KeyboardInterrupt,
    uamqp.errors.AMQPError,
    compat.TimeoutException)
):
     return  # Exception due to being closed by the load balance thread
```

How does this look to you? @annatisch , @lmazuel , @bryevdv